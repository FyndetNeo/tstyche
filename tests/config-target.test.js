import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const isNumberTestText = `import { expect, test } from "tstyche";
test("is number?", () => {
  expect<number>().type.toBeNumber();
});
`;

const tsconfig = {
  extends: "../tsconfig.json",
  include: ["**/*"],
};

const fixture = "config-target";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("'--target' command line option", () => {
  test("when single target is specified", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--target", "4.8"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when multiple targets are specified", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--target", "4.8,5.3.2,current"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when 'current' tag is specified", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--target", "current"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when 'target' configuration file option is specified", async () => {
    const config = {
      target: ["4.8", "current"],
    };

    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--target", "current"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified before the option", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["isNumber", "--target", "current"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when search string is specified after the option", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["--target", "current", "isString"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(normalizeOutput(stderr)).toBe("");

    expect(exitCode).toBe(0);
  });
});

describe("'target' configuration file option", () => {
  test("when single target is specified", async () => {
    const config = {
      target: ["4.8"],
    };

    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when multiple targets are specified", async () => {
    const config = {
      target: ["4.8", "5.3.2", "current"],
    };

    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when 'current' tag is specified", async () => {
    const config = {
      target: ["current"],
    };

    await writeFixture(fixture, {
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr } = await spawnTyche(fixture);

    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });
});
