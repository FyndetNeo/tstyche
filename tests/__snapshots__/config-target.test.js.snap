// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'--target' command line option when multiple targets are specified: stdout 1`] = `
"adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/config-target/.store/4.8.4
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

adds TypeScript 5.3.2 to <<cwd>>/tests/__fixtures__/config-target/.store/5.3.2
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

Targets:    3 passed, 3 total
Test files: 3 passed, 3 total
Tests:      3 passed, 3 total
Assertions: 3 passed, 3 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--target' command line option when search string is provided after the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/isString.tst.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran test files matching 'isString'.
"
`;

exports[`'--target' command line option when search string is provided before the option: stdout 1`] = `
"uses TypeScript <<version>>

pass ./__typetests__/isNumber.tst.ts
  + is number?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran test files matching 'isNumber'.
"
`;

exports[`'--target' command line option when single target is specified: stdout 1`] = `
"adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/config-target/.store/4.8.4
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'target' configuration file option when multiple targets are specified: stdout 1`] = `
"adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/config-target/.store/4.8.4
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

adds TypeScript 5.3.2 to <<cwd>>/tests/__fixtures__/config-target/.store/5.3.2
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

Targets:    3 passed, 3 total
Test files: 3 passed, 3 total
Tests:      3 passed, 3 total
Assertions: 3 passed, 3 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'target' configuration file option when single target is specified: stdout 1`] = `
"adds TypeScript 4.8.4 to <<cwd>>/tests/__fixtures__/config-target/.store/4.8.4
uses TypeScript <<version>> with ./tsconfig.json

pass ./__typetests__/isString.tst.ts
  + is string?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
